scapy/contrib/isotp/isotp_scan.py

```diff
@@ -0,0 +1,477 @@
+                  noise_ids=None,  # type: Optional[List[int]]
+                  sniff_time=0.1,  # type: float
+                  extended_can_id=False,  # type: bool
+                  verbose=False  # type: bool
+                  ):  # type: (...) -> Dict[int, Tuple[Packet, int]]
+    """Scan with ISOTP extended addresses and return dictionary of detections
+
+    If an answer-packet found -> slow scan with
+    single packages with extended ID 0 - 255
+    found_packets = Dictionary with Send-to-ID
+    as key and a tuple (received packet, Recv_ID)
+
+    :param sock: socket for can interface
+    :param scan_range: hexadecimal range of IDs to scan. Default is 0x0 - 0x7ff
+    :param scan_block_size: count of packets send at once
+    :param extended_scan_range: range to search for extended ISOTP addresses
+    :param noise_ids: list of packet IDs which will not be tested during scan
+    :param sniff_time: time the scan waits for isotp flow control responses
+                       after sending a first frame
+    :param extended_can_id: Send extended can frames
+    :param verbose: displays information during scan
+    :return: Dictionary with all found packets
+    """
+    return_values = dict()  # type: Dict[int, Tuple[Packet, int]]
+    scan_block_size = scan_block_size or 1
+
+    for value in scan_range:
+        if noise_ids and value in noise_ids:
+            continue
+
+        pkt = get_isotp_packet(
+            value, extended=True, extended_can_id=extended_can_id)
+        id_list = []  # type: List[int]
+        r = list(extended_scan_range)
+        for ext_isotp_id in range(r[0], r[-1], scan_block_size):
+            sock.sniff(prn=lambda p: get_isotp_fc(ext_isotp_id, id_list,
+                                                  noise_ids, True, p,
+                                                  verbose),
+                       timeout=sniff_time * 3,
+                       started_callback=lambda: send_multiple_ext(
+                           sock, ext_isotp_id, pkt, scan_block_size))
+            # sleep to prevent flooding
+            time.sleep(sniff_time)
+
+        # remove duplicate IDs
+        id_list = list(set(id_list))
+        for ext_isotp_id in id_list:
+            for ext_id in range(max(ext_isotp_id - 2, 0),
+                                min(ext_isotp_id + scan_block_size + 2, 256)):
+                pkt.extended_address = ext_id
+                full_id = (value << 8) + ext_id
+                sock.sniff(prn=lambda pkt: get_isotp_fc(full_id,
+                                                        return_values,
+                                                        noise_ids, True,
+                                                        pkt, verbose),
+                           timeout=sniff_time * 2,
+                           started_callback=lambda: sock.send(pkt))
+
+    return return_values
+
+
+def ISOTPScan(sock,  # type: SuperSocket
```

Comment:
Functions that start with caps are misleading if you ask me

Reply:
fixed!